src/analysis.py

```diff
@@ -198,79 +215,129 @@ def train_models(train_df, models, param_grid=None, output_dir=""):
         ]
     )
 
-    #Fixing year levels for ordinal encoding
-    year_levels = set(train_df['year'].unique())
-    year_levels = list(year_levels)
-    #Cutting out NaN value
-    year_levels = year_levels[1:]
-
-    ordinal_transformer = Pipeline(
+    binary_transformer = Pipeline(
         steps=[
-            ("imputer", SimpleImputer(strategy='most_frequent')),
-            ("ordinal", OrdinalEncoder(categories=[year_levels], dtype=float))
+            ("imputer", SimpleImputer(strategy='constant')),
+            ("onehot", OneHotEncoder(handle_unknown='ignore'))
```

Comment:
For binary_transformer, would you prefer to use the code below from lecture note?
`
OneHotEncoder(handle_unknown='ignore', drop='if_binary', dtype=int)
`

Reply:
So I've tried it and it's very weird... It doesn't cause any errors when we are doing the model training but it gives me some weird sklearn package errors when doing hyperparameter optimization. I thought we want consistently and it doesn't seem to have an impact on accuracy so keeping it as is. Now here's where it gets even more weirder... if I take away imputation in the model training no issue, but when I do it for hyperparameter optimization it causes a weird error "Input contains NaN" and doesn't work. With this is in mind I am going to keep imputation in both.

Reply:
For `has_last_name`, our feature script must give 0 or 1, not null values for this case, as I remember we do not have any null names. You may need strategy='most_frequent', I am not sure what constant is filling (it may not be 0 or 1, and then becomes non-binary and throws error)

Reply:
When I try 'most_frequent' or 'constant' I am getting the following error. I don't think we need imputation given there will be a value assigned from feature engineering.

ValueError: SimpleImputer does not support data with dtype bool. Please provide either a numeric array (with a floating point or integer dtype) or categorical data represented either as an array with integer dtype or an array of string values with an object dtype.

Reply:
When I try 'most_frequent' or 'constant' I am getting the following error. I don't think we need imputation given there will be a value assigned from feature engineering.

ValueError: SimpleImputer does not support data with dtype bool. Please provide either a numeric array (with a floating point or integer dtype) or categorical data represented either as an array with integer dtype or an array of string values with an object dtype.

Reply:
But good news - now using the specific OHE for binary is working.